src/metaxy/provenance/ibis.py

```diff
@@ -45,6 +50,18 @@ def __init__(
         super().__init__(plan)
         self.hash_functions: dict[HashAlgorithm, IbisHashFn] = hash_functions
 
+    @staticmethod
+    def _ensure_ibis_table(df: FrameT) -> ibis.expr.types.Table:
+        """Return the underlying Ibis table, ensuring the frame is Ibis-backed."""
+        import ibis.expr.types
+
+        native = df.to_native()
+        if not isinstance(native, ibis.expr.types.Table):
```

Comment:
I'd check df.implementation before the .to_native() call instead. It's more lightweight.